apis/duck/cached.go

```diff
@@ -38,35 +38,63 @@ var _ InformerFactory = (*CachedInformerFactory)(nil)
 // Get implements InformerFactory.
 func (cif *CachedInformerFactory) Get(gvr schema.GroupVersionResource) (cache.SharedIndexInformer, cache.GenericLister, error) {
 	cif.m.Lock()
+
 	if cif.cache == nil {
-		cif.cache = make(map[schema.GroupVersionResource]*result)
+		cif.cache = make(map[schema.GroupVersionResource]*informerCache)
 	}
-	elt, ok := cif.cache[gvr]
+
+	ic, ok := cif.cache[gvr]
 	if !ok {
-		elt = &result{}
-		elt.init = func() {
-			elt.inf, elt.lister, elt.err = cif.Delegate.Get(gvr)
+		ic = &informerCache{}
+		ic.init = func() {
+			ic.Lock()
+			defer ic.Unlock()
+
+			// double-checked lock to ensure we call the Delegate
+			// only once even if multiple goroutines end up inside
+			// init() simultaneously
+			if ic.hasInformer() {
+				return
+			}
+
+			ic.inf, ic.lister, ic.err = cif.Delegate.Get(gvr)
```

Comment:
I'm wondering if we need to store error, since AFAIK you never return the cached error?

Reply:
Probably not. The signature of `init` could also be changed to `func() error` an, if an error occurs, `Get()` would simply return `nil, nil, err`.
I can open a follow up PR if you like that better.